### PR TITLE
[RHCLOUD-35039] Generate replication event for v1 Group add / remove principal endpoints

### PR DESCRIPTION
**Pull request by lpichler**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-35039

## Description of Intent of Change(s)
This PR generated replication event for adding and removing principals.  Replication event is written in outbox table.

- reorganize code to pull out API call from transaction
- lock group and principals records
- generate replication event for added principals and service accounts

TODOs:
- [x] tests
- [x] removal principals from groups

## Local Testing
```
POST http://localhost:8000/api/rbac/v1/groups/5f067122-0e23-49ad-a604-4c91c5fe071e/principals/
x-rh-identity: 
Content-Type: application/json

{
  "principals": [
    {
      "username": "user_name"
    },
    {
      "type": "service-account",
      "clientId": "<uuid>"
    }
  ]
}
```

Example of replication event:
```
{
  "relations_to_add": [
    {
      "resource": {
        "type": "group",
        "id": "5f067122-0e23-49ad-a604-4c91c5fe071e"
      },
      "relation": "member",
      "subject": {
        "type": "user",
        "id": "98c39729-1920-4fcf-ad6d-fce4a23432eb"
      }
    },
    {
      "resource": {
        "type": "group",
        "id": "5f067122-0e23-49ad-a604-4c91c5fe071e"
      },
      "relation": "member",
      "subject": {
        "type": "user",
        "id": "cbf1e515-80ed-4683-a6c1-ac10710473d2"
      }
    }
  ],
  "relations_to_remove": []
}
```

SQL log  add principals and services accounts

https://privatebin.corp.redhat.com/?f8551a6f459c36d9#3iXYJHnq264rpeYjAJLV4QW1GCGReoQ2bkZ5FHkhKM7Q

SQL logs remove principals and service accounts 

https://privatebin.corp.redhat.com/?13873ff898734d29#128ERA1UVHUNtV4q2gtSqCdKSnmyLPXBtjZQChy4SSgY

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/group/view.py
Comment (alechenninger) [inline rbac/management/group/view.py:683]: I'm actually not sure we need to lock the group?

The main thing we're worried about is the outbox being out of order.

In the role case, the outbox event was _dependent_ upon all events before it, because it referenced existing IDs that could have been changed by the very previous / concurrent tx.

In the group principal case, the IDs are always the same – group, user & service account ID – regardless of previous state. So I think we should be able to get away without locking the group too in this case.

For example, there can definitely be race conditions but I don't think they "matter" in that, though they may appear odd in some cases, they still result in consistent relations. For example:

1. tx1 add user1 start
2. tx2 add user1 start
3. tx1 add user commit – outbox add user1
4. tx3 remove user start
5. tx3 remove user commit – outbox remove user1
6. tx2 add user commit – outbox add user1 

This would result in a user added even though we just removed them, but the rbac db & relations would still be consistent about this. It's the same outcome as if tx2 just started a "little" bit later and was after tx3 instead of interleaving.

Locking would change the order (tx2 would block until tx1 committed, tx3 would block until tx2 committed, and therefore we'd end up with the user removed instead of added), but again this is only the difference in slight timings. It doesn't change the consistency of the data.

The main thing I think is just that any group changes MUST be committed with the associated outbox event wholly representing those changes in the same transaction.

Can you think of a case where we need to lock the group?

File: rbac/management/group/relation_api_dual_write_group_handler.py
Comment (alechenninger) [inline rbac/management/group/relation_api_dual_write_group_handler.py:99]: May I ask why the additional query here?

I'm suspicious it would cause race conditions (and require us to lock the group) although I haven't proven that yet. But regardless of that, it maybe seems unnecessary?

If another tx removed the users from the group, that tx should commit and write its own replication event.

I'm hoping (perhaps naively) that group principal add/remove can be very simple, essentially 3 sql statements:

For a req to add principals

1. insert principals & insert group fks from principals in request
2. insert outbox event w/ principals in request
3. commit

And likewise for remove.

Comment (lpichler) [reply rbac/management/group/relation_api_dual_write_group_handler.py:99]: Group is locked but query may be seems unnecessary. It was about what id of principal and service account we will have in relation. 

Request payload contains username of principal and clientID service account. 

so I will keep UUID as id in relation - eventually we can change it later.

Anyway any IDs could be passed as they are available from additions - it just need to be extracted and passed here.

I will remove query and pass UUID it here.

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:436]: I was assuming this: What if principal record is delete during this transaction by other transaction  ? Is this case ok ?

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:683]: Thanks @alechenninger for explaining! 

Mainly I was thinking if group is deleted during current transaction - do you think it is reason to lock group ?

1. tx1 add user1 start
2. tx2 delete group
3. tx1 sent data to outbox table with group which doesn't exist.

This was my reasoning

Comment (alechenninger) [reply rbac/management/group/view.py:690]: Yeah great question. The answer to this is basically the same for the principal case. I think you're talking about this scenario:

Tx1 | Tx2 | Result | Input | Output | Tuple state
-- | -- | -- | -- | -- | --
check if g1 exists |   | ok | g1 | exists |  
add user1 g1 |   | ok | u1, g1 |   |  
add outbox event |   | ok |   |   |  
  | check if g1 exists | ok |   | exists |  
  | remove g1 | ok | g1 |   |  
  | select for update role bindings w/ g1 | ok | g1 | role binding list |  
  | update / remove role bindings | ok | new role binding state |   |  
  | add outbox event | ok |   |   |  
  | commit | ok |   |   | ok
commit |   | ok |   |   | not ok! outbox event adds orphaned group#member@user tuple

I don't think that is actually what would happen. The reason is because of foreign key constraints. 

```
ALTER TABLE "management_group_principals" ADD CONSTRAINT "management_group_pri_group_id_0c1b3b8d_fk_managemen" FOREIGN KEY ("group_id") REFERENCES "management_group" ("id") DEFERRABLE INITIALLY DEFERRED; (params None)

ALTER TABLE "management_group_principals" ADD CONSTRAINT "management_group_pri_principal_id_47b7a6f8_fk_managemen" FOREIGN KEY ("principal_id") REFERENCES "management_principal" ("id") DEFERRABLE INITIALLY DEFERRED; (params None)
```

What should happen is this:

Tx1 | Tx2 | Result | Input | Output | Tuple state
-- | -- | -- | -- | -- | --
check if g1 exists |   | ok | g1 | exists |  
add user1 g1 |   | ok | u1, g1 |   |  
add outbox event |   | ok |   |   |  
  | check if g1 exists | ok |   | exists |  
  | remove g1 | ok | g1 |   |  
  | select for update role bindings w/ g1 | ok | g1 | role binding list |  
  | update / remove role bindings | ok | new role binding state |   |  
  | add outbox event | ok |   |   |  
  | commit | ok |   |   | ok
commit |   | rollback - fk constraint violation |   |   | ok – no event committed (rollback)

When tx1 goes to commit, it shouldn't succeed because of the foreign key constraints b/w the group and the principal. Might be worth confirming this by manually running sql against two concurrent sessions in the db but I think that's what should happen. In that case, as long as the outbox event and the user add are in the same tx (atomic block), the outbox event actually never gets written because the user addition never gets written, either, all without locking.

That said, we can consider locking the group _anyway_ since it's not too expensive and might save us in case there is a scenario we aren't thinking of. But I think we're okay.

For principals, it's the same thing. If a concurrent tx removes the principal and commits, and then we try to refer to that principal with another commit, that will rollback b/c the principal no longer exists.

**Comment by lpichler** [comment]
Currently PR is done.

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:694]: Thanks ! I will leave here locking and principals here in that case.

Comment (alechenninger) [reply rbac/management/group/view.py:694]: Not sure I follow – what I am describing is why the lock is not necessary. The first table isn't actually what would happen, I was clarifying the difference between I think what you were expecting vs what I think would actually happen.

It's not a huge deal to leave the locks but shouldn't be necessary. Are you saying you want to leave it just in case?

File: rbac/management/group/view.py
Comment (alechenninger) [inline rbac/management/group/view.py:481]: Looks like the type signature is wrong here 

```suggestion
    ) -> Tuple[Group, List[Principal]]:
```

File: rbac/management/group/view.py
Comment (alechenninger) [inline rbac/management/group/view.py:1146]: I _think_ this additional atomic block is unnecessary now

File: rbac/management/group/view.py
Comment (alechenninger) [inline rbac/management/group/view.py:1156]: One note around the notification handlers: previously these happend _after_ the writes committed. So the semantics you had prior to this were that you might not send a message (due to failure) even if you successfully committed something.

With this change, that's different. Now it's the other way around: because we haven't committed yet, we send the message regardless of whether our commit rolls back, which means you might instead send a message when you shouldn't have.

So it's whether you want to err on the side of sending when you shouldn't, vs not sending when you should've. I'm not sure which is better–maybe even this is an improvement–but just something to keep in mind for edge cases.

File: rbac/management/group/view.py
Comment (alechenninger) [inline rbac/management/group/view.py:1162]: Do we need to port anything else from `get_group_queryset` here?

There are various filters that take into account user access and query params. I'm not sure if those are relevant here, but we did end up porting some of that over for the Role view at least.

This is maybe one cost of locking – if we don't really need to, we could consider keeping the same query and we wouldn't have to run a second query, either.

**Review by alechenninger — COMMENTED** [review]
Nothing major, just a few questions / small things. Looking good!

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:694]: yes I understand what did you describe but according to statement and I also confirmed it with 2 sessions.
in TX1, it tries to insert a group ID that no longer exists because it was deleted by transaction TX2, resulting in an error due to a constraint violation

I intended to leave it here because of this sentence:

> That said, we can consider locking the group anyway since it's not too expensive and might save us in case there is a scenario we aren't thinking of. But I think we're okay.

Now I understand that you prefer not to use locking for groups and principals, 
so I will remove it. 

NOTE: With locks, the only difference is that TX2 will wait for TX1, but most likely it will fail because TX2 is attempting to delete a group that now has a reference in the management_group_principals table, which was added by TX1.

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:1146]: good catch, thanks!

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:1156]: This is good point to consider - in addition if sending into kafka will fail - I guess it will rollback whole transaction. Which is probably fine as it means that debezium will also not work :) 

(I will keep as it is - as I see role dual write has same change   )

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:1162]: yes I am going to remove locking as I mentioned in previous commit.

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:480]: Thanks, done 👍

**Comment by lpichler** [comment]
Thanks @alechenninger, I added your suggestions 👍

File: rbac/management/group/view.py
Comment (alechenninger) [reply rbac/management/group/view.py]: Thanks for clarifying and confirming locally!

I guess one another small downside to locking is that if there are other transactions trying to write to the group in some way, those will block while the group is locked (even if they are unrelated to dual write, e.g. group name or something like that). A small thing, though!

File: rbac/management/group/view.py
Comment (alechenninger) [reply rbac/management/group/view.py:1154]: That's a good point. It means if Kafka is down, more functionality is impacted (severely!). I suppose that is another reason to consider moving notifications to use the outbox pattern, too. Not only would it be atomic, but it would also be more available (and probably reduce latencies on API calls, too).

We may want to reconsider the role changes that brought that into the TX as well, but I'll open an issue for it and we can discuss with the team.

**Comment by alechenninger** [comment]
@lpichler Cool I think this is good to 🚢 then! I think I will wait for #1199 to be reviewed & merged before merging this one to keep that review isolated.

File: rbac/management/group/relation_api_dual_write_group_handler.py
Comment (alechenninger) [inline rbac/management/group/relation_api_dual_write_group_handler.py:43]: Since these are both just `List[Principal]` what do you think about combining the parameters? It makes it a little easier to use (e.g. in tests) because I don't have to explicitly sort users vs services ahead of time, and its not actualy needed as far as I can tell since we combine them into one below (`for principal in self.principal_usernames + self.service_accounts`).

**Review by alechenninger — COMMENTED** [review]
Small minor things but with or without LGTM!

(Note: I pushed up some more tests and updates to fixtures in test_dual_write – just a way to test more scenarios. In the future we may be able to use some of those cases to also test against full stack relations + spicedb checks.)

File: rbac/management/group/view.py
Comment (lpichler) [reply rbac/management/group/view.py:694]: you are right, it was not intent (probably leftover after I removed locking groups), good catch!

File: rbac/management/group/relation_api_dual_write_group_handler.py
Comment (lpichler) [reply rbac/management/group/relation_api_dual_write_group_handler.py]: I did not combine this initially because I wasn't sure if we would need separate IDs for the service account and principals but as you mentioned it seems unnecessary for now. Anyway If it becomes necessary later, we can handle it in the dual write handler.

I updated also your tests and I removed service account creation as well - but if you intended to keep it for later use  we can revert/remove last commit.

**Comment by lpichler** [comment]
@alechenninger I have addressed your latest suggestions, thanks!

File: rbac/management/group/relation_api_dual_write_group_handler.py
Comment (alechenninger) [reply rbac/management/group/relation_api_dual_write_group_handler.py]: K cool! Let's keep the service account creation support in the test fixture – that is still a use case we want to test explicitly I think, regardless of the implementation.

As for IDs, a fair point. I am thinking both will end up using "user_id" field being added in #1179 though, so we might be okay.

**Review by alechenninger — APPROVED** [review]
Ship it!

**Pull request merged**
